File: 0328-odd-even-linked-list/0328-odd-even-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

/*
  Traverse through linked list and have a variable to track even nodes (e.g. counter)
  
     1    ->    2    ->    3    ->    4    ->   5  ->   6
    odd        evenTail      nextOdd   nextEven
     1    ->    3    ->    2    ->    4    ->    5  ->   6
               odd                  evenTail
*/
var oddEvenList = function(head) {
  if (!head) return null;
  if (!head.next) return head;
  
  let odd = head, evenHead = head.next, evenTail = head.next, nextOdd, nextEven;
  
  while (evenTail && evenTail.next) {
    nextOdd = evenTail.next;
    nextEven = nextOdd.next;
    
    odd.next = nextOdd;
    evenTail.next = nextEven;
    odd = nextOdd;
    evenTail = nextEven;
  }
  
  odd.next = evenHead;
  return head;
  
};

File: 0328-odd-even-linked-list/0328-odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
          return None
        if not head.next:
          return head
        
        odd = head
        even_head = even_tail = head.next
        
        # Link all odd & even nodes together
        
        while even_tail and even_tail.next:
          next_odd = even_tail.next
          next_even = next_odd.next
          
          odd.next = next_odd
          even_tail.next = next_even
          
          odd = next_odd
          even_tail = next_even
        
        # Finally, link tail of odd to head of even nodes (even_head)
        odd.next = even_head
        
        return head